src/Microsoft.AspNetCore.Blazor/Routing/Router.cs

```diff
@@ -96,31 +65,32 @@ private string StringUntilAny(string str, char[] chars)
                 : str.Substring(0, firstIndex);
         }
 
-        private Type FindComponentTypeInAssemblyOrReferences(Assembly assembly, string typeName)
-            => assembly.GetType(typeName, throwOnError: false, ignoreCase: true)
-            ?? assembly.GetReferencedAssemblies()
-                .Select(Assembly.Load)
-                .Select(referencedAssembly => FindComponentTypeInAssemblyOrReferences(referencedAssembly, typeName))
-                .FirstOrDefault();
-
-        protected virtual void Render(RenderTreeBuilder builder, Type matchedComponentType)
+        protected virtual void Render(RenderTreeBuilder builder, Type handler, IDictionary<string, string> parameters)
         {
             builder.OpenComponent(0, typeof(LayoutDisplay));
-            builder.AddAttribute(1, nameof(LayoutDisplay.Page), matchedComponentType);
+            builder.AddAttribute(1, nameof(LayoutDisplay.Page), handler);
+            builder.AddAttribute(2, nameof(LayoutDisplay.PageParameters), parameters);
             builder.CloseComponent();
         }
 
         private void Refresh()
         {
             var locationPath = UriHelper.ToBaseRelativePath(_baseUriPrefix, _locationAbsolute);
-            var matchedComponentType = GetComponentTypeForPath(locationPath);
-            if (!typeof(IComponent).IsAssignableFrom(matchedComponentType))
+            locationPath = StringUntilAny(locationPath, _queryOrHashStartChar);
+            var context = new RouteContext(locationPath);
+            Routes.Route(context);
```

Comment:
Could lines 80 and 81 be moved into a `protected virtual` method called `GetComponentTypeForPath` that by default returns `context.Handler`? Then people can still subclass `Router` if they want, and put in completely custom route matching logic if they need it. I don't think we need to pass the `RouteTable` into this method, as that's internal - we should only pass the original `locationPath` (i.e., before truncating it at `_queryOrHashStartChar`).

Or do you think there's an argument that people *shouldn't* implement custom route matching logic?